File: DailyChallenges/CPP/LeetCode_23.cpp
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b)->bool { return a->val > b->val; };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> pq(cmp);
        for (auto l: lists) { if (l) pq.push(l); }
        ListNode* ans = nullptr;
        while (!pq.empty()){
            auto l = pq.top(); pq.pop();
            if (l){
                if (l->next) pq.push(l->next);
                l->next = ans;
                ans = l;
            }
        }
        
        function<ListNode* (ListNode*, ListNode*)> rev = [&](ListNode* a, ListNode* b) {
            if (!a) return b; else{
                auto _a = a->next;
                a->next = b; 
                return rev(_a, a);
            }
        };
        
        return rev(ans, nullptr);
    }
};

File: DailyChallenges/Scala/LeetCode_23.scala
object Solution {
    def mergeKLists(lists: Array[ListNode]): ListNode = {
        // convert to a big list
        def loop(xl: List[Int] = List.empty[Int]): List[Int] = {
            val _xl = lists.zipWithIndex.foldLeft(xl)((l, p) => {
                if (p._1 != null){
                    lists(p._2) = p._1.next
                    p._1.x :: l
                }else l
            })
        
            if (_xl.length == xl.length) _xl // done
            else loop(_xl)
        }
        
        val xl = loop().sorted
        if (xl.isEmpty) null else{
            val head: ListNode = new ListNode(xl.head)

            xl.tail.foldLeft(head)((node, x) => {
                val _next = new ListNode(x)
                node.next = _next
                _next
            })

            head
        }
    }
}
